examples/direct_fidelity_estimation.py

```diff
@@ -52,65 +50,129 @@ def compute_characteristic_function(circuit: cirq.Circuit,
     return trace, prob
 
 
+async def estimate_characteristic_function(
+        circuit: cirq.Circuit, P_i: Tuple[cirq.Gate, ...],
+        qubits: List[cirq.Qid], simulator: cirq.DensityMatrixSimulator,
+        samples_per_term: int):
+    n_qubits = len(P_i)
+
+    pauli_string = cirq.PauliString(dict(zip(qubits, P_i)))
+    qubit_map = dict(zip(qubits, range(n_qubits)))
+
+    p = cirq.PauliSumCollector(circuit=circuit,
+                               observable=pauli_string,
+                               samples_per_term=samples_per_term)
+
+    completion = p.collect_async(sampler=simulator)
+    assert await completion is None
+
+    sigma_i = p.estimated_energy()
+    assert np.isclose(sigma_i.imag, 0.0, atol=1e-6)
+    sigma_i = sigma_i.real
+
+    return sigma_i
+
+
 def direct_fidelity_estimation(circuit: cirq.Circuit, qubits: List[cirq.Qid],
-                               noise: cirq.NoiseModel, n_trials: int):
+                               noise: cirq.NoiseModel, n_trials: int,
+                               samples_per_term: int):
     # n_trials is upper-case N in https://arxiv.org/pdf/1104.3835.pdf
 
     # Number of qubits, lower-case n in https://arxiv.org/pdf/1104.3835.pdf
     n_qubits = len(qubits)
+    d = 2**n_qubits
 
     # Computes for every \hat{P_i} of https://arxiv.org/pdf/1104.3835.pdf,
     # estimate rho_i and Pr(i). We then collect tuples (rho_i, Pr(i), \hat{Pi})
     # inside the variable 'pauli_traces'.
     pauli_traces = []
+
+    simulator = cirq.DensityMatrixSimulator()
+    # rho in https://arxiv.org/pdf/1104.3835.pdf
+    clean_density_matrix = cast(
+        cirq.DensityMatrixTrialResult,
+        simulator.simulate(circuit)).final_density_matrix
+
+    # TODO(tonybruguier): Sample the Pauli states more efficenitly when the
```

Comment:
prefer opening a github issue and linking that here instead of a single person's name